server/src/main/java/io/spine/server/delivery/ShardedStream.java

```diff
@@ -0,0 +1,361 @@
+import static com.google.common.base.Preconditions.checkNotNull;
+import static io.spine.server.delivery.ShardedStream.GenericParameter.MESSAGE_CLASS;
+import static io.spine.util.Exceptions.newIllegalStateException;
+
+/**
+ * The stream of messages of a particular type sent for the processing to a specific shard.
+ *
+ * <p>Uses a {@linkplain io.spine.server.transport.MessageChannel MessageChannel} as an
+ * underlying transport.
+ *
+ * <p>A typical example is a stream of {@code Event}s, being sent from the write-side
+ * to a particular shard of read-side {@code Projection}s.
+ *
+ * @param <I> the type of identifiers of message targets, such as entities
+ * @param <M> the type of messages, which are being sharded and sent via this stream
+ * @param <E> the type of envelopes into which the messages are packed
+ *
+ * @author Alex Tymchenko
+ */
+public abstract class ShardedStream<I, M extends Message, E extends MessageEnvelope<?, M, ?>>
+        implements AutoCloseable {
+
+    /**
+     * A name of the bounded context, within which the sharded stream exists.
+     */
+    private final BoundedContextName boundedContextName;
+
+    /**
+     * A key, which defines a shard.
+     */
+    private final ShardingKey key;
+
+    /**
+     * A tag which defines a type
+     */
+    private final DeliveryTag<E> tag;
+    private final Class<I> targetIdClass;
+    private final Subscriber subscriber;
+    private final Publisher publisher;
+    private final ExternalMessageObserver channelObserver;
+
+    /**
+     * A lazily-intialized converted for the sharded messages.
+     */
+    @Nullable
+    private ShardedMessageConverter<I, M, E> converter;
+
+    ShardedStream(AbstractBuilder<I, E, ?, ? extends ShardedStream> builder) {
+        this.boundedContextName = builder.boundedContextName;
+        this.key = builder.key;
+        this.tag = builder.tag;
+        this.targetIdClass = builder.targetIdClass;
+        final Class<E> envelopeCls = getEnvelopeClass();
+        final ChannelId channelId = toChannelId(builder.boundedContextName, key, envelopeCls);
+        this.subscriber = builder.transportFactory.createSubscriber(channelId);
+        this.publisher = builder.transportFactory.createPublisher(channelId);
+
+        this.channelObserver = new ExternalMessageObserver(builder.consumer);
+        this.subscriber.addObserver(channelObserver);
+    }
+
+    @SuppressWarnings("unchecked")  // Ensured by the generic type definition.
+    private Class<E> getEnvelopeClass() {
+        return (Class<E>) MESSAGE_CLASS.getArgumentIn(getClass());
+    }
+
+    private Class<I> getTargetIdClass() {
+        return targetIdClass;
+    }
+
+    private ShardedMessageConverter<I, M, E> converter() {
+        if(converter == null) {
+            converter = newConverter();
+        }
+        return converter;
+    }
+
+    protected abstract ShardedMessageConverter<I, M, E> newConverter();
+
+    public ShardingKey getKey() {
```

Comment:
Please document this method, and `getTag()`. Assuming the general character of the words `key` and `tag`, we need docs.